File: src/main/java/com/github/cannor147/model/rgb/RGBColor.java
package com.github.cannor147.model.rgb;

import lombok.*;

import java.awt.*;

@Getter
@ToString
@EqualsAndHashCode
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class RGBColor implements RGB {
    public static RGBColor of(int red, int green, int blue) {
        return new RGBColor(red & 0xff, green & 0xff, blue & 0xff);
    }

    public static RGBColor fromInt(int rgbInt) {
        return of(rgbInt >> 16, rgbInt >> 8, rgbInt);
    }

    private final int red;
    private final int green;
    private final int blue;

    @Override
    public RGB add(RGB other) {
        return of(this.red + other.getRed(), this.green + other.getGreen(), this.blue + other.getBlue());
    }

    @Override
    public RGB subtract(RGB other) {
        return of(this.red - other.getRed(), this.green - other.getGreen(), this.blue - other.getBlue());
    }

    @Override
    public RGB multiply(double proportion) {
        return of((int) (this.red * proportion), (int) (this.green * proportion), (int) (this.blue * proportion));
    }

    @Override
    public RGBColor asColor() {
        return this;
    }

    public Color toColor() {
        return new Color(red, green, blue);
    }

    public int toInt() {
        return toColor().getRGB();
    }
}

File: src/main/java/com/github/cannor147/painter/RGBColor.kt
package com.github.cannor147.painter

import java.awt.Color

data class RGBColor internal constructor(
    override val red: Int,
    override val green: Int,
    override val blue: Int,
) : RGB {
    override fun add(other: RGB): RGB {
        return of(red + other.red, green + other.green, blue + other.blue)
    }

    override fun subtract(other: RGB): RGB {
        return of(red - other.red, green - other.green, blue - other.blue)
    }

    override fun multiply(proportion: Double): RGB {
        return of((red * proportion).toInt(), (green * proportion).toInt(), (blue * proportion).toInt())
    }

    override fun asColor(): RGBColor {
        return this
    }

    @Suppress("MemberVisibilityCanBePrivate")
    fun toColor(): Color {
        return Color(red, green, blue)
    }

    fun toInt(): Int {
        return toColor().rgb
    }

    companion object {
        @JvmStatic
        fun of(red: Int, green: Int, blue: Int): RGBColor {
            return RGBColor(red and 0xff, green and 0xff, blue and 0xff)
        }

        fun fromInt(rgbInt: Int): RGBColor {
            return of(rgbInt shr 16, rgbInt shr 8, rgbInt)
        }
    }
}